README.md
# bootstrap-progressbar - 0.8.5

`bootstrap-progressbar` is a [jQuery](http://jquery.com) plugin which extends the basic [twitter-bootstrap](https://github.com/twbs/bootstrap) progressbar. It provides the ability to animate the progressbar by adding Javascript in combination with the preexisting css transitions. Additionally you can display the current progress information in the bar or get the value via callback.


## What's new in v0.8.0 - v0.8.5?

* Add SCSS support
* Add bootstrap 3.3.4 styles
* Add bootstrap 3.3.2 styles
* Add bootstrap 3.3.1 styles
* Add bootstrap 3.3.0 styles
* Add bootstrap 3.2.0 styles
* Extend `amount_format` hook for min value
* Update demo page
* Remove unused travis badge
* Fix #27 switch `aria-valuetransitionsgoal` to `data-transitionsgoal`
* Fix #35 wrong amount calculation for `aria-valuemin` != 0
* Fix #33 pass `$this` to `update` and `done` callback
* Fix #39 revert low percentage styles in bootstrap 3.2

> __Note__: this version replaces the not official supported `aria-valuetransitionsgoal` with `data-transitionsgoal` which may affect (or brick) your progressbars.

> __Note__: Bootstrap's [low percentage styles](http://getbootstrap.com/components/#progress-low-percentages) (introduced in 3.2.0) causes flickering under some circumstances (e.g. slow transitions). To preserve correct progressbar functionality + appearance this bs feature is reverted in v0.8.3. (and reverted by bootstrap in 3.3.0)

> __Note__: Stacked progressbars are not supported by now.


## Demo

* http://www.minddust.com/project/bootstrap-progressbar/demo/


## Installation

* Download the latest release: [v0.8.5](https://github.com/minddust/bootstrap-progressbar/archive/v0.8.5.zip)
* Clone the repository: `git clone git@github.com:minddust/bootstrap-progressbar.git`
* Install with [Bower](http://bower.io): `bower install bootstrap-progressbar`


## Usage

1. include `bootstrap-progressbar.js`

    ```html
    <script type="text/javascript" src="bootstrap-progressbar.js"></script>
    ```

2. activate `bootstrap-progressbar` functionality on progressbars of your choice:


    ```javascript
    $('.progress .bar').progressbar();           // bootstrap 2
    $('.progress .progress-bar').progressbar();  // bootstrap 3
    ```

3. set the `aria` attribute and __remove__ the `width` style attribute (alternatively you can set it to 0)

    1. `data-transitiongoal`

        ```html
        <div class="progress">
            <div class="progress-bar" data-transitiongoal="75"></div>
        </div>
        ```

    2. `aria-valuemin` (default: 0) and `aria-valuemax` (default: 100)

        ```html
        <div class="progress">
            <div class="progress-bar" data-transitiongoal="75" aria-valuemin="-1337" aria-valuemax="9000"></div>
        </div>
        ```


## Usage Extended

* Do I need the additional style file?

    * for the horizontal bar with no or filled text: __NO__
    * for any vertical bars or the horizontal bar with centered text or right aligned bar: __YES__

        less:

        ```html
        <link rel="stylesheet/less" type="text/css" href="bootstrap-progressbar.less">
        <script src="less.js" type="text/javascript"></script>
        ```

        or precompile it yourself. (see: [3rd party compilers](http://lesscss.org/usage/#third-party-compilers-node-js-compilers))

        scss:
        * need to precompile it yourself (see: [libsass](http://sass-lang.com/libsass))

        css:

        ```html
        <link rel="stylesheet" type="text/css" href="bootstrap-progressbar.css">
        ```

        you can find your needed css, less or scss files inside the specific subdirectories.

* Multiple trigger

    You can trigger progressbar as much as you want. Just change your attribute(s) and trigger `.progressbar()` again. All settings made before will be kept.


## Settings

### default values

```javascript
Progressbar.defaults = {
    transition_delay: 300,
    refresh_speed: 50,
    display_text: 'none',
    use_percentage: true,
    percent_format: function(percent) { return percent + '%'; },
    amount_format: function(amount_part, amount_total) { return amount_part + ' / ' + amount_total; },
    update: $.noop,
    done: $.noop,
    fail: $.noop
};
```

### transition_delay

Is the time in milliseconds until the animation starts.

This could be useful to delay the start on the initial page load like:

```javascript
$(document).ready(function() {
    $('.progress .progress-bar').progressbar({
        transition_delay: 1500
    });
});
```

### refresh_speed

Is the time in milliseconds which will elapse between every text refresh, `aria-valuenow` attribute update and `update` callback call.

### display_text

Determines if and where to display text on the progressbar. Possible options:

* `none` __no text__
* `fill` __text on filled bar__
* `center` __text on center__ *(this mode changes css / html due to styling requirements)*

### use_percentage

If text will be displayed - this option determines whether to show the percentage value or the amount.

So if `use_percentage` is false and `aria-valuemin` and `aria-valuemax` are not set (or to `0` and `100`) the value will be the same but `amount_format` will be used to format the result.

Example:

`<div class="progress-bar" data-transitiongoal="75">`

with `use_percentage: true` is the final text: `75%`

with `use_percentage: false` is the final text: `75 / 100`

### percent_format

Is a function which returns the text format for progressbar with `use_percentage: true`.

It takes 1 argument which is the current percent value.

### amount_format

Is a function which returns the text format for progressbar with `use_percentage: false`.

It takes 3 argument which are the current-, the max- and the min-amount.

### update

Is a callback function which will be called while the progressbar is transitioning ;)

Depends on `refresh_speed`.

It takes 2 argument which is the current percent value and a reference to the attached progressbar element.

### done

Is a callback function which will be called when the transition process is done.

It takes 1 argument which is a reference to the attached progressbar element.

### fail

Is a callback function which will be called when an error occurs.

It takes 1 argument which is the error message.


## Customisation

### alignment
* to use a horizontal progressbar which is align to the right you have to add `right` to the `progress` element.

   ```html
   <div class="progress right">
   ```
* to use a vertical progressbar you have to add `vertical` to the `progress` element.

   ```html
   <div class="progress vertical">
   ```
* to use a vertical progressbar which is align to the bottom you have to add `vertical` and `bottom` to the `progress` element.

   ```html
   <div class="progress vertical bottom">
   ```

### animation

to change the animation itself you have to overwrite either less or css

1. horizontal
    * less

        ```css
        .progress .bar {
            .transition(width 2s ease-in-out);
        }
        ```
    * scss

        ```css
        .progress.vertical .progress-bar {
            @include transition(width 2s ease-in-out);
        }
        ```
    * css

        ```css
        .progress .bar {
            -webkit-transition: width 2s ease-in-out;
            -moz-transition: width 2s ease-in-out;
            -ms-transition: width 2s ease-in-out;
            -o-transition: width 2s ease-in-out;
            transition: width 2s ease-in-out;
        }
        ```

1. vertical
    * less

        ```css
        .progress.vertical .bar {
            .transition(height 2s ease-in-out);
        }
        ```
    * scss

        ```css
        .progress.vertical .bar {
            @include transition(height 2s ease-in-out);
        }
        ```
    * css

        ```css
        .progress.vertical .bar {
            -webkit-transition: height 2s ease-in-out;
            -moz-transition: height 2s ease-in-out;
            -ms-transition: height 2s ease-in-out;
            -o-transition: height 2s ease-in-out;
            transition: height 2s ease-in-out;
        }
        ```

## Known Problems

* Looks like iOS Safari is flooring the width of the transition. So if you want to display text with a correct value you have to use a full bar width **greater or equal 100px**.

## Copyright and license

Copyright 2013-2015 Stephan Groß, under [MIT license](https://github.com/minddust/bootstrap-progressbar/blob/master/LICENSE).

Want to appreciate my work? [minddust at Gittip](https://www.gittip.com/minddust/)


CHANGELOG.md
# Changelog

## 0.8.5 (2015-04-04)

* Add bootstrap 3.3.1-3.3.4 styles
* Add SCSS support (request #46)
* Fix #42 wrong version name

## 0.8.4 (2014-11-05)

* Add bootstrap 3.3.0 styles

## 0.8.3 (2014-08-08)

* Fix #39 revert low percentage styles in bootstrap 3.2

## 0.8.2 (2014-08-02)

* Fix #33 pass `$this` to `update` and `done` callback

## 0.8.1 (2014-08-01)

* Fix #35 wrong amount calculation for `aria-valuemin` != 0

## 0.8.0 (2014-08-01)

* Add bootstrap 3.2.0 styles
* Extend `amount_format` hook for min value
* Update demo page
* Remove unused travis badge
* Fix #27 switch `aria-valuetransitionsgoal` to `data-transitionsgoal`

### 0.7.1 (2014-03-05)

* Add bootstrap 3.1.1 styles
* Switch to Gulp
* Remove bs dotfiles
* Fix #30 compile errors with old markup
* Fix #29 wrong style generation
* Fix #28 element creation which will brick with `django-compressor`

### 0.7.0 (2014-02-11)

* Add bootstrap 3.0.0-3.1.0 styles
* Fix #22 wrong vertical styles for bootstrap 2.x
* Fix #18 missing css requirement documentation

### 0.6.0 (2013-08-16)

* Add bootstrap 3 support
* Add `noConflict` fallback
* Add striped styles
* Add custom string formatting
* Rename `filled` option to `fill`
* Switch `display_text` option dict to string setting
* Switch from `data` to `aria` attributes
* Switch license from Apache2 to MIT
* Fix some minor style issues

### 0.5.0 (2012-08-28)

* Add vertical progressbars
* Add right alignment
* Move styles to css/less
* Fix some code formatting
* Fix different font sizes

### 0.4.6 (2012-06-27)

* Fix flickering on centered text on multiple trigger (due to last fix)

### 0.4.5 (2012-06-25)

* Fix stacking bug on centered text on multiple trigger

### 0.4.4 (2012-06-13)

* Fix caching bug with data method
* Add multitrigger support

### 0.4.3 (2012-06-11)

* Switch from jQuery's `*.attr('data-*')` to `*.data('*')`

### 0.4.2 (2012-06-07)

* Start changelog
* Switch to github markdown


Gulpfile.js
var gulp = require('gulp');
var util = require('gulp-util');

var clean = require('gulp-clean');
var header = require('gulp-header');
var less = require('gulp-less');
var rename = require('gulp-rename');
var uglify = require('gulp-uglify');
var minifyCss = require('gulp-minify-css');

var pkg = require('./package.json');
var currentYear = util.date(new Date(), 'yyyy');

var paths = {
    scripts: [
        './bootstrap-progressbar.js'
    ],
    styles: [
        './resources/bootstrap-progressbar-2.0.0.less',
        './resources/bootstrap-progressbar-2.0.1.less',
        './resources/bootstrap-progressbar-2.0.2.less',
        './resources/bootstrap-progressbar-2.0.3.less',
        './resources/bootstrap-progressbar-2.0.4.less',
        './resources/bootstrap-progressbar-2.1.0.less',
        './resources/bootstrap-progressbar-2.1.1.less',
        './resources/bootstrap-progressbar-2.2.0.less',
        './resources/bootstrap-progressbar-2.2.1.less',
        './resources/bootstrap-progressbar-2.2.2.less',
        './resources/bootstrap-progressbar-2.3.0.less',
        './resources/bootstrap-progressbar-2.3.1.less',
        './resources/bootstrap-progressbar-2.3.2.less',
        './resources/bootstrap-progressbar-3.0.0-rc1.less',
        './resources/bootstrap-progressbar-3.0.0-rc2.less',
        './resources/bootstrap-progressbar-3.0.0.less',
        './resources/bootstrap-progressbar-3.0.1.less',
        './resources/bootstrap-progressbar-3.0.2.less',
        './resources/bootstrap-progressbar-3.0.3.less',
        './resources/bootstrap-progressbar-3.1.0.less',
        './resources/bootstrap-progressbar-3.1.1.less',
        './resources/bootstrap-progressbar-3.2.0.less',
        './resources/bootstrap-progressbar-3.3.0.less',
        './resources/bootstrap-progressbar-3.3.1.less',
        './resources/bootstrap-progressbar-3.3.2.less',
        './resources/bootstrap-progressbar-3.3.4.less',
    ]
};

var banner = '/*! <%= pkg.name %> v<%= pkg.version %> | Copyright (c) 2012-<%= currentYear %> <%= pkg.author %> | <%= pkg.license %> license | <%= pkg.homepage %> */\n';

gulp.task('scripts', function() {
    return gulp.src(paths.scripts)
        .pipe(uglify())
        .pipe(header(banner, {pkg: pkg, currentYear: currentYear}))
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('.'));
});

gulp.task('styles', function() {
    return gulp.src(paths.styles)
        .pipe(less())
        .pipe(header(banner, {pkg: pkg, currentYear: currentYear}))
        .pipe(gulp.dest('./css'));
});

gulp.task('styles-min', function() {
    return gulp.src(paths.styles)
        .pipe(less())
        .pipe(minifyCss())
        .pipe(header(banner, {pkg: pkg, currentYear: currentYear}))
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('./css'));
});

gulp.task('clean', function() {
    return gulp.src('./css', {read: false})
        .pipe(clean());
});

gulp.task('default', ['clean', 'scripts', 'styles', 'styles-min']);


Gruntfile.js
module.exports = function(grunt) {

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        banner: '/*! <%= pkg.name %> v<%= pkg.version %> | Copyright (c) 2012-<%= grunt.template.today("yyyy") %> <%= pkg.author %> | MIT License | minddust.com */\n',

        clean: {
            css: ['css']
        },

        uglify: {
            options: {
                banner: '<%= banner %>'
            },
            build: {
                src: '<%= pkg.name %>.js',
                dest: '<%= pkg.name %>.min.js'
            }
        },

        recess: {
            options: {
                compile: true
            },
            bs_2_0_0:     { src: ['resources/bootstrap-progressbar-2.0.0.less'], dest: 'css/bootstrap-progressbar-2.0.0.css' },
            bs_2_0_0_min: { src: ['resources/bootstrap-progressbar-2.0.0.less'], dest: 'css/bootstrap-progressbar-2.0.0.min.css', options: { compress: true } },
            bs_2_0_1:     { src: ['resources/bootstrap-progressbar-2.0.1.less'], dest: 'css/bootstrap-progressbar-2.0.1.css' },
            bs_2_0_1_min: { src: ['resources/bootstrap-progressbar-2.0.1.less'], dest: 'css/bootstrap-progressbar-2.0.1.min.css', options: { compress: true } },
            bs_2_0_2:     { src: ['resources/bootstrap-progressbar-2.0.2.less'], dest: 'css/bootstrap-progressbar-2.0.2.css' },
            bs_2_0_2_min: { src: ['resources/bootstrap-progressbar-2.0.2.less'], dest: 'css/bootstrap-progressbar-2.0.2.min.css', options: { compress: true } },
            bs_2_0_3:     { src: ['resources/bootstrap-progressbar-2.0.3.less'], dest: 'css/bootstrap-progressbar-2.0.3.css' },
            bs_2_0_3_min: { src: ['resources/bootstrap-progressbar-2.0.3.less'], dest: 'css/bootstrap-progressbar-2.0.3.min.css', options: { compress: true } },
            bs_2_0_4:     { src: ['resources/bootstrap-progressbar-2.0.4.less'], dest: 'css/bootstrap-progressbar-2.0.4.css' },
            bs_2_0_4_min: { src: ['resources/bootstrap-progressbar-2.0.4.less'], dest: 'css/bootstrap-progressbar-2.0.4.min.css', options: { compress: true } },
            bs_2_1_0:     { src: ['resources/bootstrap-progressbar-2.1.0.less'], dest: 'css/bootstrap-progressbar-2.1.0.css' },
            bs_2_1_0_min: { src: ['resources/bootstrap-progressbar-2.1.0.less'], dest: 'css/bootstrap-progressbar-2.1.0.min.css', options: { compress: true } },
            bs_2_1_1:     { src: ['resources/bootstrap-progressbar-2.1.1.less'], dest: 'css/bootstrap-progressbar-2.1.1.css' },
            bs_2_1_1_min: { src: ['resources/bootstrap-progressbar-2.1.1.less'], dest: 'css/bootstrap-progressbar-2.1.1.min.css', options: { compress: true } },
            bs_2_2_0:     { src: ['resources/bootstrap-progressbar-2.2.0.less'], dest: 'css/bootstrap-progressbar-2.2.0.css' },
            bs_2_2_0_min: { src: ['resources/bootstrap-progressbar-2.2.0.less'], dest: 'css/bootstrap-progressbar-2.2.0.min.css', options: { compress: true } },
            bs_2_2_1:     { src: ['resources/bootstrap-progressbar-2.2.1.less'], dest: 'css/bootstrap-progressbar-2.2.1.css' },
            bs_2_2_1_min: { src: ['resources/bootstrap-progressbar-2.2.1.less'], dest: 'css/bootstrap-progressbar-2.2.1.min.css', options: { compress: true } },
            bs_2_2_2:     { src: ['resources/bootstrap-progressbar-2.2.2.less'], dest: 'css/bootstrap-progressbar-2.2.2.css' },
            bs_2_2_2_min: { src: ['resources/bootstrap-progressbar-2.2.2.less'], dest: 'css/bootstrap-progressbar-2.2.2.min.css', options: { compress: true } },
            bs_2_3_0:     { src: ['resources/bootstrap-progressbar-2.3.0.less'], dest: 'css/bootstrap-progressbar-2.3.0.css' },
            bs_2_3_0_min: { src: ['resources/bootstrap-progressbar-2.3.0.less'], dest: 'css/bootstrap-progressbar-2.3.0.min.css', options: { compress: true } },
            bs_2_3_1:     { src: ['resources/bootstrap-progressbar-2.3.1.less'], dest: 'css/bootstrap-progressbar-2.3.1.css' },
            bs_2_3_1_min: { src: ['resources/bootstrap-progressbar-2.3.1.less'], dest: 'css/bootstrap-progressbar-2.3.1.min.css', options: { compress: true } },
            bs_2_3_2:     { src: ['resources/bootstrap-progressbar-2.3.2.less'], dest: 'css/bootstrap-progressbar-2.3.2.css' },
            bs_2_3_2_min: { src: ['resources/bootstrap-progressbar-2.3.2.less'], dest: 'css/bootstrap-progressbar-2.3.2.min.css', options: { compress: true } },
            bs_3_0_0_rc1:     { src: ['resources/bootstrap-progressbar-3.0.0-rc1.less'], dest: 'css/bootstrap-progressbar-3.0.0-rc1.css' },
            bs_3_0_0_rc1_min: { src: ['resources/bootstrap-progressbar-3.0.0-rc1.less'], dest: 'css/bootstrap-progressbar-3.0.0-rc1.min.css', options: { compress: true } },
            bs_3_0_0_rc2:     { src: ['resources/bootstrap-progressbar-3.0.0-rc2.less'], dest: 'css/bootstrap-progressbar-3.0.0-rc2.css' },
            bs_3_0_0_rc2_min: { src: ['resources/bootstrap-progressbar-3.0.0-rc2.less'], dest: 'css/bootstrap-progressbar-3.0.0-rc2.min.css', options: { compress: true } },
            bs_3_0_0:     { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.0.css' },
            bs_3_0_0_min: { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.0.min.css', options: { compress: true } },
            bs_3_0_1:     { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.1.css' },
            bs_3_0_1_min: { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.1.min.css', options: { compress: true } },
            bs_3_0_2:     { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.2.css' },
            bs_3_0_2_min: { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.2.min.css', options: { compress: true } },
            bs_3_0_3:     { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.3.css' },
            bs_3_0_3_min: { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.0.3.min.css', options: { compress: true } },
            bs_3_1_0:     { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.1.0.css' },
            bs_3_1_0_min: { src: ['resources/bootstrap-progressbar-3.0.0.less'], dest: 'css/bootstrap-progressbar-3.1.0.min.css', options: { compress: true } }
        },

        file_append: {
            default_options: {
                files: {
                    'css/bootstrap-progressbar-2.0.0.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.0.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.1.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.1.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.2.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.2.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.3.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.3.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.4.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.0.4.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.1.0.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.1.0.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.1.1.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.1.1.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.2.0.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.2.0.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.2.1.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.2.1.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.2.2.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.2.2.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.3.0.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.3.0.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.3.1.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.3.1.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.3.2.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-2.3.2.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.0-rc1.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.0-rc1.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.0-rc2.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.0-rc2.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.0.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.0.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.1.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.1.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.2.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.2.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.3.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.0.3.min.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.1.0.css': { prepend: '<%= banner %>' },
                    'css/bootstrap-progressbar-3.1.0.min.css': { prepend: '<%= banner %>' }
                }
            }
        }
    });


    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-recess');
    grunt.loadNpmTasks('grunt-file-append');

    grunt.registerTask('test', []);  // TODO: include tests task
    grunt.registerTask('dist-js', ['uglify']);
    grunt.registerTask('dist-css', ['recess', 'file_append']);
    grunt.registerTask('dist', ['clean', 'dist-css', 'dist-js']);

    grunt.registerTask('default', ['test', 'dist']);
};


tasks/converter.rb
# coding: utf-8
# Based on convert script from vwall/compass-twitter-bootstrap gem.
# https://github.com/vwall/compass-twitter-bootstrap/blob/master/build/convert.rb
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this work except in compliance with the License.
# You may obtain a copy of the License in the LICENSE file, or at:
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'open-uri'
require 'json'
require 'strscan'
require 'forwardable'
require 'term/ansicolor'
require 'fileutils'

require_relative 'less_conversion'
require_relative 'logger'

class Converter
  extend Forwardable
  include LessConversion

  def initialize()
    @logger     = Logger.new
    @save_to    = { scss: 'scss' }
  end

  def_delegators :@logger, :log, :log_status, :log_processing, :log_transform, :log_file_info, :log_processed, :log_http_get_file, :log_http_get_files, :silence_log

  def process_bootstrap_progressbar
    log_status "Convert LESS to Sass"
    @save_to.each { |_, v| FileUtils.mkdir_p(v) }
    process_stylesheet_assets
  end

  def save_file(path, content, mode='w')
    dir = File.dirname(path)
    FileUtils.mkdir_p(dir) unless File.directory?(dir)
    File.open(path, mode) { |file| file.write(content) }
  end
end


bootstrap-progressbar.js
/*!
 * bootstrap-progressbar v0.8.1 by @minddust
 * Copyright (c) 2012-2014 Stephan Groß
 *
 * http://www.minddust.com/project/bootstrap-progressbar/
 *
 * Licensed under the MIT license:
 * http://www.opensource.org/licenses/MIT
 */
(function($) {

    'use strict';

    // PROGRESSBAR CLASS DEFINITION
    // ============================

    var Progressbar = function(element, options) {
        this.$element = $(element);
        this.options = $.extend({}, Progressbar.defaults, options);
    };

    Progressbar.defaults = {
        transition_delay: 300,
        refresh_speed: 50,
        display_text: 'none',
        use_percentage: true,
        percent_format: function(percent) { return percent + '%'; },
        amount_format: function(amount_part, amount_max, amount_min) { return amount_part + ' / ' + amount_max; },
        update: $.noop,
        done: $.noop,
        fail: $.noop
    };

    Progressbar.prototype.transition = function() {
        var $this = this.$element;
        var $parent = $this.parent();
        var $back_text = this.$back_text;
        var $front_text = this.$front_text;
        var options = this.options;
        var data_transitiongoal = parseInt($this.attr('data-transitiongoal'));
        var aria_valuemin = parseInt($this.attr('aria-valuemin')) || 0;
        var aria_valuemax = parseInt($this.attr('aria-valuemax')) || 100;
        var is_vertical = $parent.hasClass('vertical');
        var update = options.update && typeof options.update === 'function' ? options.update : Progressbar.defaults.update;
        var done = options.done && typeof options.done === 'function' ? options.done : Progressbar.defaults.done;
        var fail = options.fail && typeof options.fail === 'function' ? options.fail : Progressbar.defaults.fail;

        if (isNaN(data_transitiongoal)) {
            fail('data-transitiongoal not set');
            return;
        }
        var percentage = Math.round(100 * (data_transitiongoal - aria_valuemin) / (aria_valuemax - aria_valuemin));

        if (options.display_text === 'center' && !$back_text && !$front_text) {
            this.$back_text = $back_text = $('<span>').addClass('progressbar-back-text').prependTo($parent);
            this.$front_text = $front_text = $('<span>').addClass('progressbar-front-text').prependTo($this);

            var parent_size;

            if (is_vertical) {
                parent_size = $parent.css('height');
                $back_text.css({height: parent_size, 'line-height': parent_size});
                $front_text.css({height: parent_size, 'line-height': parent_size});

                $(window).resize(function() {
                    parent_size = $parent.css('height');
                    $back_text.css({height: parent_size, 'line-height': parent_size});
                    $front_text.css({height: parent_size, 'line-height': parent_size});
                }); // normal resizing would brick the structure because width is in px
            }
            else {
                parent_size = $parent.css('width');
                $front_text.css({width: parent_size});

                $(window).resize(function() {
                    parent_size = $parent.css('width');
                    $front_text.css({width: parent_size});
                }); // normal resizing would brick the structure because width is in px
            }
        }

        setTimeout(function() {
            var current_percentage;
            var current_value;
            var this_size;
            var parent_size;
            var text;

            if (is_vertical) {
                $this.css('height', percentage + '%');
            }
            else {
                $this.css('width', percentage + '%');
            }

            var progress = setInterval(function() {
                if (is_vertical) {
                    this_size = $this.height();
                    parent_size = $parent.height();
                }
                else {
                    this_size = $this.width();
                    parent_size = $parent.width();
                }

                current_percentage = Math.round(100 * this_size / parent_size);
                current_value = Math.round(aria_valuemin + this_size / parent_size * (aria_valuemax - aria_valuemin));

                if (current_percentage >= percentage) {
                    current_percentage = percentage;
                    current_value = data_transitiongoal;
                    done();
                    clearInterval(progress);
                }

                if (options.display_text !== 'none') {
                    text = options.use_percentage ? options.percent_format(current_percentage) : options.amount_format(current_value, aria_valuemax, aria_valuemin);

                    if (options.display_text === 'fill') {
                        $this.text(text);
                    }
                    else if (options.display_text === 'center') {
                        $back_text.text(text);
                        $front_text.text(text);
                    }
                }
                $this.attr('aria-valuenow', current_value);

                update(current_percentage, $this);
            }, options.refresh_speed);
        }, options.transition_delay);
    };


    // PROGRESSBAR PLUGIN DEFINITION
    // =============================

    var old = $.fn.progressbar;

    $.fn.progressbar = function(option) {
        return this.each(function () {
            var $this = $(this);
            var data = $this.data('bs.progressbar');
            var options = typeof option === 'object' && option;

            if (!data) {
                $this.data('bs.progressbar', (data = new Progressbar(this, options)));
            }
            data.transition();
        });
    };

    $.fn.progressbar.Constructor = Progressbar;


    // PROGRESSBAR NO CONFLICT
    // =======================

    $.fn.progressbar.noConflict = function () {
        $.fn.progressbar = old;
        return this;
    };

})(window.jQuery);
